protocol/codecs/base_scalar.go

```diff
@@ -209,10 +209,15 @@ func (c *Bytes) Type() reflect.Type {
 // Decode []byte.
 func (c *Bytes) Decode(msg *buff.Message, out reflect.Value) {
 	b := msg.PopBytes()
-	// todo use out's memory if it has enough allocated :thinking:
-	o := make([]byte, len(b))
-	copy(o, b)
-	out.SetBytes(o)
+	p := (*[]byte)(unsafe.Pointer(out.UnsafeAddr()))
+
+	if cap(*p) >= len(b) {
+		*p = (*p)[:len(b)]
+	} else {
+		*p = make([]byte, len(b))
```

Comment:
What are the scenarios when `out` might not have enough memory pre-allocated?

Reply:
Out has zero bytes allocated.

```go
var out []byte
con.QueryOneJson(ctx, query, &out)
```

The same thing happens if a struct with a slice field is allocated this way. The slice gets zero bytes.